Break the machine controller's dependency on Google Actuator

krousey opened the pull request:
This changes the machine controller library to take the machine actuator as a parameter and not construct it based on a flag an factory function. This allows a higher level package to provide an implementation to the library without the library having an import dependency on it.

As an example, I wrote a ext-apiserver/cloud/cmd/gce-machine-controller that passes in the GCE machine actuator. It's less than 40 lines of code. There's nothing special about the actuator's implementation being in-tree here.

This does modify some of the auto-generated code. It changes the way a machine controller is constructed in `ext-apiserver/pkg/controller/machine/zz_generated.api.register.go` to take an actuator as a parameter. It also changes `ext-apiserver/pkg/controller/zz_generated.api.register.go` to remove the machine controller from the set of all controllers that will be pulled into generic controller-manager binary.

This change is not complete, but enough to get the idea across. Specifically tests and the gce deployer need updating, and a new docker image for the new controller need to be created.

Are we OK with a change like this?

@kubernetes/kube-deploy-reviewers

Commit:
Break the machine controller's dependency on Google Actuator

krousey commented:
/hold

krousey commented:
Just to verify...

Before:
```console
> go list -json k8s.io/kube-deploy/ext-apiserver/pkg/controller | grep -iE '(cloud/google)|(google.golang.org/api)'           
		"k8s.io/kube-deploy/ext-apiserver/cloud/google",
		"k8s.io/kube-deploy/ext-apiserver/cloud/google/config",
		"k8s.io/kube-deploy/ext-apiserver/cloud/google/gceproviderconfig",
		"k8s.io/kube-deploy/ext-apiserver/cloud/google/gceproviderconfig/v1alpha1",
		"k8s.io/kube-deploy/ext-apiserver/vendor/google.golang.org/api/compute/v1",
		"k8s.io/kube-deploy/ext-apiserver/vendor/google.golang.org/api/gensupport",
		"k8s.io/kube-deploy/ext-apiserver/vendor/google.golang.org/api/googleapi",
		"k8s.io/kube-deploy/ext-apiserver/vendor/google.golang.org/api/googleapi/internal/uritemplates",
```

After:
```console
> git checkout break_cloud
Previous HEAD position was 1f4cd706 Merge pull request #647 from medinatiger/cleanup
Switched to branch 'break_cloud'
Your branch is up to date with 'origin/break_cloud'.
> go list -json k8s.io/kube-deploy/ext-apiserver/pkg/controller | grep -iE '(cloud/google)|(google.golang.org/api)'
>
```

krousey commented:
cc @medinatiger @kcoronado

craigtracey commented:
cc: @kris-nova 

At first glance, yes, I think the "flipping of the logic" wherein instead of the library calling the cloud-specific actuators, the cloud-specific controllers call the library is the better approach. 

I do not think this makes new controller development burdensome...in fact makes it quite a bit more flexible.

rsdcastro commented:
+1

This also makes sense after chatting with @cheftako and @calebamiles about the possibility of pulling the provider code into the Cloud Provider repos they are working on.

p0lyn0mial commented:
I guess changing `apiserver-builder` in a way that would allow for custom initialization is not an option ? The simplest thing we could do is to export `MachineController.controller` field. Extending `ControllerInit ` interface would be the best. Especially that it is defined as follows:
```
// ControllerInit new controllers should implement this.  It is more flexible in
// allowing additional options to be passed in
type ControllerInit interface {
	Init(args ControllerInitArguments)
}
```

I think other users would benefit from change like this as well :) ?

krousey commented:
@p0lyn0mial I thought about that approach when I saw it, but unfortunately the generated code still constructs that argument explicitly and doesn't really allow for customization without modifying the generated code.

```go
	} else if i, ok := ci.(sharedinformers.ControllerInit); ok {	
		i.Init(&sharedinformers.ControllerInitArgumentsImpl{si, config, c.LookupAndReconcile})	
	}
```

We could change the API server builder, but I'm wondering if this code falls into the "generate once and never again" category unlike the conversion and deep copy code.

medinatiger commented:
What's the real benefits here? It does make writing deployment tool easier to mix & match between cloud provider and machine controller implementation. I like this part. However, it's now outside controller manager with rest of the controllers (machineset, cluster, machinedeployment, etc). It brings its own challenge in terms of how to deploy it. 

Another option is to add some config option in controller manager to choose whether or not to include each controller type. If desired, the user can choose to use third-party machine controller and don't enable the built-in machine controller in controller manager.

krousey commented:
@medinatiger I think the controller being separated is a benefit. This way, I only have to worry about publishing my own MachineController and the sig can publish standard MachineSet or MachineDeployment controllers in a controller manager container that anyone can consume. I don't really see it being that much harder to deploy this. We're already deploying an API server, ETCD cluster, and one controller manager. What's another controller to that?

I agree that as we start adding more controllers (deployment, etc..), we need to make it easy to disable certain controllers via config if people want to provide their own. But I still think separating the machine controller out as a separate binary all-together is the right choice.

p0lyn0mial commented:
> We could change the API server builder, but I'm wondering if this code falls into the "generate once and never again" category unlike the conversion and deep copy code.

@krousey not sure about that but  I'm just afraid that it is not practical to maintain a custom code inside a generated file in a long run. No one will remember about putting a custom code in the right place:(

Nevertheless if we decide to follow that path and we are certain that we want to have a custom initialisation I think I could try to implement such a feature in `apiserver-builder` at some point.

roberthbailey commented:
Needs rebase.

krousey commented:
OK I've fixed all the tests, fixed the gcp-deployer to work with this new controller, and since I wrote a docker and make file for it, I wrote some for the API server and controller manager as well. This should be good for final review.

@p0lyn0mial I think we are going to move forward with this with the thought (and hope) that the code we have to frequently regenerate doesn't overlap with this. If you want to work to upstream a fix to the builder that allows us to do this better, I'm all for it.

roberthbailey commented on cluster-api/gcp-deployer/deploy/deploy.go:
```diff
@@ -52,7 +54,7 @@ func NewDeployer(provider string, configPath string) *deployer {
 		}
 	}
-	md, err := newMachineDeployer(provider, token)
+	md, err := google.NewMachineActuator(token, nil)
```
this variable shouldn't be `md` any longer since you aren't assigning it a deployer.

krousey replied:
Ha. It never was a machine deployer. Just a weird name for a factory function.

krousey replied:
Done.

roberthbailey commented on cluster-api/cmd/controller-manager/Makefile:
```diff
@@ -0,0 +1,32 @@
+	docker build -t "$(PREFIX)/$(NAME):$(TAG)" -f ./Dockerfile ../..
+
+push: image
+	docker push "$(PREFIX)/$(NAME):$(TAG)"
```
maybe it's just me but i can only seem to push to gcr using gcloud docker rather than just docker.

rsdcastro replied:
@roberthbailey what is the issue you run into? Perhaps some auth issue?

It seems we should not require gcloud if this is not a GCP specific code.

krousey replied:
This maybe got lost in my email reply, but `gcloud beta auth configure-docker` will install gcloud as a credential store for docker. This allows `docker push` to work with gcr.io.

roberthbailey commented on cluster-api/cloud/google/config/configtemplate.go:
```diff
@@ -110,8 +110,26 @@ spec:
             memory: 30Mi
-      - name: controller
-        image: {{ .Image }}
+      - name: controller-manager
```
so does this controller manager have the generic machineset controller (and will later include the generic machinedeployment one) in it and the gce-machine-controller (below) have just the gce machine controller? Is that the reason to add a third container?

krousey replied:
Yes. This way people can re-use our (cluster-api's) API server, MachineSet controller, MachineDeployment controller, and just replace the machine controller with whatever they need.

krousey commented:
gcloud beta auth configure-docker is what you're looking for.

On Thu, Mar 22, 2018, 07:03 Rodrigo De Castro <notifications@github.com>
wrote:

> *@rsdcastro* commented on this pull request.
> ------------------------------
>
> In cluster-api/cmd/controller-manager/Makefile
> <https://github.com/kubernetes/kube-deploy/pull/649#discussion_r176430288>
> :
>
> > +# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> +# See the License for the specific language governing permissions and
> +# limitations under the License.
> +
> +.PHONY: image push dev_image dev_push
> +
> +PREFIX = gcr.io/k8s-cluster-api
> +DEV_PREFIX ?= gcr.io/$(shell gcloud config get-value project)
> +NAME = controller-manager
> +TAG = 0.0.1
> +
> +image:
> +	docker build -t "$(PREFIX)/$(NAME):$(TAG)" -f ./Dockerfile ../..
> +
> +push: image
> +	docker push "$(PREFIX)/$(NAME):$(TAG)"
>
> @roberthbailey <https://github.com/roberthbailey> what is the issue you
> run into? Perhaps some auth issue?
>
> It seems we should not require gcloud if this is not a GCP specific code.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kube-deploy/pull/649#discussion_r176430288>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJlm7rTvRdKarlJWCCarjBxZHubo-7Lks5tg68XgaJpZM4SrQX6>
> .
>

karan commented:
This requires documentation updates as well right? Please add those here.

p0lyn0mial commented on cluster-api/cloud/google/cmd/gce-machine-controller/main.go:
```diff
@@ -0,0 +1,67 @@
+
+	shutdown := make(chan struct{})
+	si := sharedinformers.NewSharedInformers(config, shutdown)
+	c := machine.NewMachineController(config, si, actuator)
```
do you think that adding a comment what to do in the event of compilation error would be helpful :) ?

krousey replied:
Done.

p0lyn0mial commented on cluster-api/pkg/controller/machine/zz_generated.api.register.go:
```diff
@@ -44,25 +44,15 @@ type MachineController struct {
 
 // NewController returns a new MachineController for responding to Machine events
-func NewMachineController(config *rest.Config, si *sharedinformers.SharedInformers) *MachineController {
+func NewMachineController(config *rest.Config, si *sharedinformers.SharedInformers, actuator Actuator) *MachineController {
```
Did you consider exposing `MachineController.controller` field instead of passing an `actuator` as a parameter ? In that case fixing failing build would be easier and in the future we could inject more custom code directly to `MachineImplementation`

krousey replied:
I much prefer an API contract where I get a fully functioning object back from `New`. Having to further initialize the object after constructing just seems wrong. If anything, we could perhaps pass in the controller instead of the actuator.

Exposing the `.controller` field allows for partial initialization. One case where this is bad is when "cluster-api/pkg/controller/zz_generated.api.register.go" is regenerated to add a new controller. If the machine controller is added back there (which I don't know if it does), it would compile fine, but it would run an uninitialized `MachineController` that would nil panic as soon as it tried to access the `.controller` field. I would much rather deal with the confusion of compilation errors for now.

krousey commented:
@karan I checked, and nothing I've done changes anything that is documented that I could find. The existing documentation doesn't cover how to build and deploy images, how to write your own, or how to use those ENV vars to do dev testing with GCE. I agree that these things should be documented, but I'd rather not undertake that in this PR.

roberthbailey commented:
/lgtm
